File: Gateway 1.2/gift1.cpp
/*
TASK: gift1
LANG: C++
*/

#include <vector>
#include <iostream>
#include <unordered_map>
#include <string>
#include <fstream>
#include <vector>

using namespace std;

int main()
{
    fstream fin("gift1.in", ios_base::in);
    fstream fout("gift1.out", ios_base::out);
    int num;
    fin >> num;
    string name;
    vector<string> names;
    unordered_map<string, int> m1;
    for (int i = 0; i < num; i++) {
        fin >> name;
        names.push_back(name);
        m1[name] = 0;
    }

    while (fin >> name) {
        int amount;
        fin >> amount >> num;
        int dist = 0;
        if (num != 0) {
            dist = amount / num;
            int left = amount % num;
            m1[name] -= (amount-left);
        }
        else {
            continue;
        }

        for (int i = 0; i < num; i++) {
            fin >> name;
            m1[name] += dist;
        }
    }

    for (auto& e : names) {
        fout << e << " " << m1[e] << endl;
    }
}

File: Gateway 1.2/gift1.java
/*
LANG: JAVA
TASK: gift1
*/

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Scanner;

public class gift1 {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(new File("gift1.in"));
        int n = sc.nextInt();
        sc.nextLine();  // need to be put extra causions here!

        String[] names = new String[n];
        for (int i = 0; i < n; i++) {
            names[i] = sc.nextLine();
        }

        // name: money, name:money, name: money
        HashMap<String, Integer> pockets = new HashMap<>();  
        for (String s : names) {
            pockets.put(s, 0);
        }

        for (int i = 0; i < n; i++) {
            String giver = sc.nextLine();
            int cashOut = sc.nextInt();
            int shareCount = sc.nextInt();
            sc.nextLine();

            if (shareCount > 0) {
                String[] receivers = new String[shareCount];
                for (int j = 0; j < shareCount; j++) {
                    receivers[j] = sc.nextLine();
                }
                int moneyForOne = cashOut / shareCount;  // 100 / 3 = 33
                int remainder = cashOut % shareCount;    // 100 % 3 = 1
                for (String receiver : receivers) {
                    int currentbalance = pockets.get(receiver);
                    currentbalance += moneyForOne;
                    pockets.put(receiver,  currentbalance );
                }
                pockets.put(giver, pockets.get(giver) - cashOut + remainder);
            }
        }

        FileWriter fw = new FileWriter("gift1.out");
        for (String s : names)
            fw.write(s + " " + pockets.get(s) + "\n");
        fw.close();
    }
}
